crates/bitwarden/src/vault/totp.rs

```diff
@@ -22,9 +34,244 @@ pub struct TotpResponse {
 /// - Steam URI
 ///
 /// Supports providing an optional time, and defaults to current system time if none is provided.
-pub async fn generate_totp(_key: String, _time: Option<DateTime<Utc>>) -> TotpResponse {
-    TotpResponse {
-        code: "000 000".to_string(),
-        period: 30,
+///
+/// Arguments:
+/// - `key` - The key to generate the TOTP code from
+/// - `time` - The time in UTC to generate the TOTP code for, defaults to current system time
+pub async fn generate_totp(key: String, time: Option<DateTime<Utc>>) -> Result<TotpResponse> {
+    let params: Totp = key.parse()?;
+
+    let time = time.unwrap_or_else(Utc::now);
+
+    let otp = params.derive_otp(time.timestamp())?;
+
+    Ok(TotpResponse {
+        code: otp,
+        period: params.period,
+    })
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Algorithm {
+    Sha1,
+    Sha256,
+    Sha512,
+    Steam,
+}
+
+impl Algorithm {
+    // Derive the HMAC hash for the given algorithm
+    fn derive_hash(&self, key: &[u8], time: &[u8]) -> Result<Vec<u8>> {
+        fn compute_digest<D: Mac>(mut digest: D, time: &[u8]) -> Vec<u8> {
+            digest.update(time);
+            digest.finalize().into_bytes().to_vec()
+        }
+
+        Ok(match self {
+            Algorithm::Sha1 => compute_digest(HmacSha1::new_from_slice(key)?, time),
+            Algorithm::Sha256 => compute_digest(HmacSha256::new_from_slice(key)?, time),
+            Algorithm::Sha512 => compute_digest(HmacSha512::new_from_slice(key)?, time),
+            Algorithm::Steam => compute_digest(HmacSha1::new_from_slice(key)?, time),
+        })
+    }
+}
+
+#[derive(Debug)]
+struct Totp {
+    algorithm: Algorithm,
+    digits: u32,
+    period: u32,
+    secret: String,
+}
+
+impl Default for Totp {
+    fn default() -> Self {
+        Self {
+            algorithm: Algorithm::Sha1,
+            digits: 6,
+            period: 30,
+            secret: "".to_string(),
+        }
+    }
+}
+
+impl Totp {
+    fn derive_otp(&self, time: i64) -> Result<String> {
+        let time = time / self.period as i64;
+
+        let secret = BASE32
```

Comment:
I think the secret should be base32 decoded in the from_str implementation, with that and the change suggested in `derive_hash`, we can make this function infallible